examples/RecordBuilder/README.md

```diff
@@ -0,0 +1,118 @@
+
+# Record Builder 
+
+This example demonstrates the Record Builder pattern in Roc. This pattern leverages the functional programming concept of applicative functors, to provide a flexible method for constructing complex types.
+
+## The Basics
+
+Let's assume we want to develop a module that supplies a type-safe yet versatile method for users to safely obtain sequential IDs. The record builder pattern can be beneficial in this scenario.
+
+1. **Opaque Type** We need an opaque type that will accumulate our state:
+```roc
+Count a := (U32, a)
+```
+This type takes a type variable `a`, and holds a tuple of a `U32` and an `a`. Here, U32 will maintain our counter's value, while a houses the function to progress our builder pattern.
+
+2. **End Goal** It's useful to visualize our desired result. The record builder pattern we're aiming for looks like:
+
+```roc
+expect
+    { foo, bar, baz } = 
+        from {
+            foo: <- inc,
+            bar: <- inc,
+            baz: <- inc,
+        } |> done
+
+    foo == 1 && bar == 2 && baz == 3
+```
+
+This generates a record with fields foo, bar, and baz, all possessing sequential U32 IDs. Note the slight deviation from the conventional record syntax, using a `: <-` instead of `:`, this is the record builder pattern syntax.
```

Comment:
I don't know if we should call Record Builder syntax a "pattern". To me, the pattern is Applicative Functors, though I'm wary of overusing that term because it sounds intimidating.
```suggestion
This generates a record with fields `foo`, `bar`, and `baz`, all possessing sequential U32 IDs. Note the slight deviation from the conventional record syntax, using a `: <-` instead of `:`, this is the Record Builder syntax.
```